docs: add postgres reference section to docs (close #4440) (#4471)
hasura/graphql-engine: docs/graphql/manual/guides/postgres/constraints.rst

.. meta::
   :description: Use Postgres constraints with Hasura
   :keywords: hasura, docs, postgres, constraints

.. _postgres_constraints:

Postgres constraints
====================

.. contents:: Table of contents
  :backlinks: none
  :depth: 2
  :local:

Introduction
------------

`Postgres constraints <https://www.postgresql.org/docs/current/ddl-constraints.html>`__ are used to define rules for fields in a database table. They ensure that
no invalid data is entered into the database.

.. note::

  For more detailed information on Postgres constraints, please refer to the `Postgres documentation <https://www.postgresql.org/docs/current/ddl-constraints.html>`__.

Postgres constraints
--------------------

There are different types of constraints that can be used with Postgres.

Primary key constraints
^^^^^^^^^^^^^^^^^^^^^^^

> A ``PRIMARY KEY`` is used to identify each specific row of a table.

Reviewer: ```suggestion
A ``PRIMARY KEY`` is used to identify each unique row of a table.
```

**Identify the author's id as the primary key of the authors table:**

.. code-block:: sql
  :emphasize-lines: 2

  CREATE TABLE authors(

>     id INT PRIMARY KEY     NOT NULL,

Reviewer: PRIMARY KEY is inherently NOT NULL, so this is redundant

    name           TEXT    NOT NULL
  );

Foreign key constraints
^^^^^^^^^^^^^^^^^^^^^^^

A foreign key constraint specifies that the values in a column must match the values appearing in a row of another table. 
Foreign key constraints are used to create relationships between tables.

**Define the author_id in the articles table as a foreign key to the id field in the authors table:**

.. code-block:: sql
  :emphasize-lines: 5

  CREATE TABLE articles(

>     id SERIAL PRIMARY KEY  NOT NULL,

Reviewer: PRIMARY KEY is inherently NOT NULL

    title          TEXT    NOT NULL,
    author_id INTEGER,

>     FOREIGN KEY (author_id) REFERENCES authors (id)

Reviewer: should probably show the `authors` table definition as well?

Author: Done

  );

Not-null constraints
^^^^^^^^^^^^^^^^^^^^

A not-null constraint allows you to specify that a column's value cannot be ``null``.

**Validate that an author's id and name cannot be null:**

.. code-block:: sql
  :emphasize-lines: 2-3

  CREATE TABLE authors(

>     id SERIAL PRIMARY KEY  NOT NULL,

Reviewer: PRIMARY KEY is inherently NOT NULL

    name           TEXT    NOT NULL
  );

Unique constraints
^^^^^^^^^^^^^^^^^^

> Unique constraints prevent database entries with a duplicate value of the respective field.

Reviewer: ```suggestion
Unique constraints prevent two rows with same value of the respective field.
```

**Validate that an author's indicator is unique:**

.. code-block:: sql
  :emphasize-lines: 4

  CREATE TABLE authors(
    id SERIAL PRIMARY KEY  NOT NULL,
    name           TEXT    NOT NULL,

>     indicator      TEXT    UNIQUE

Reviewer: why not use `email` as a more common field?

  );

Check constraints
^^^^^^^^^^^^^^^^^

> Check constraints allow you to specify a ``Boolean`` expression for a specific field. 

Reviewer: ```suggestion
Check constraints allow you to specify a ``Boolean`` expression for a table. 
```

> This Boolean expression must be satisfied (equal to ``true``) by the field value for the object to be inserted.

Reviewer: Check constraints can span multiple fields.

> This Boolean expression must be satisfied (equal to ``true``) by the field value for the object to be inserted.

Reviewer: ```suggestion
This boolean expression must be satisfied (equate to ``true``) on a row for it to be inserted into this table.
```

**Validate that an author's rating is between 1 and 10:**

.. code-block:: sql
  :emphasize-lines: 4

  CREATE TABLE authors(

>     id SERIAL PRIMARY KEY  NOT NULL,

Reviewer: PRIMARY KEY is inherently NOT NULL
